datasets/utils/secret.py

```diff
@@ -0,0 +1,213 @@
+        if self.raw_secret:
+            return self._fetch_raw_secret()
+
+    def _variable_validation(self):
+        """
+        Ensure exact one secret source variable provided.
+        Ensure key is of str type if provided
+        """
+        secret_source_variables = ["kubernetes_secret_name", "aws_secret_arn", "env_var", "raw_secret"]
+
+        secret_source_variable_used = [
+            getattr(self, variable) is not None for variable in secret_source_variables
+        ]
+
+        if secret_source_variable_used.count(True) != 1:
+            raise ValueError("Must provide exact one secret source variable!")
+
+        if self.key is not None and not isinstance(self.key, str):
+            raise ValueError("key should be an string!")
+
+    def _fetch_cluster_secret(self) -> SECRET_RETURN_TYPE:
+        # Try to fetch from cache first
+        global secret_cache
+        secret_from_cache = secret_cache["cluster_secret"].get(self.kubernetes_secret_name)
+        if secret_from_cache is not None and not self.force_reload:
+            logger.info(f"Using secret from cache {self.kubernetes_secret_name}")
+            secret_value = secret_from_cache
+        else:
+            config.load_incluster_config()
+            core_api = client.CoreV1Api()
+            namespace = get_current_namespace()
+            for attempt in Retrying(stop=stop_after_attempt(3), wait=wait_fixed(5), reraise=True):
+                with attempt:
+                    raw_secret_value = core_api.read_namespaced_secret(
+                        self.kubernetes_secret_name, namespace
+                    ).data
+
+            # Ensure we convert to Dict[str, str] type
+            secret_value = {}
+            for k, v in raw_secret_value.items():
+                # The last decode("utf-8") converts bytes type to str
+                secret_value[str(k)] = base64.b64decode(v).decode("utf-8")
+            secret_cache["cluster_secret"][self.kubernetes_secret_name] = secret_value
+
+        return secret_value[self.key] if self.key is not None else secret_value
+
+    def _fetch_aws_secret(self) -> SECRET_RETURN_TYPE:
+        global secret_cache
+        secret_from_cache = secret_cache["aws_secret"].get(self.aws_secret_arn)
+        if secret_from_cache is not None and not self.force_reload:
+            logger.info(f"Using secret from cache {self.aws_secret_arn}")
+            secret_value = secret_from_cache
+        else:
+            # We have AWS_REGION env var injected in most of our workloads
+            secrets_manager_client = boto3.Session(region_name=os.getenv("AWS_REGION", "us-west-2")).client(
+                "secretsmanager"
+            )
+
+            for attempt in Retrying(
+                retry=retry_if_exception(
+                    lambda e: e.response["Error"]["Code"] == "InternalServiceErrorException"
+                ),
+                stop=stop_after_attempt(3),
+                wait=wait_fixed(5),
```

Comment:
Minor, do we want to expose these stop/wait retry numbers? Obviously would have the same defaults as those sound good but I wonder if in 2 years someone wants to change these defaults having these parameterized now makes it trivial 😅